cylc/flow/ws_data_mgr.py

```diff
@@ -479,6 +540,62 @@ def update_workflow(self):
         # Replace the originals (atomic update, for access from other threads).
         self.workflow = workflow
 
+    def _prune_points(self, point_strings):
+        """Remove old proxies and edges by cycle point."""
+        if not point_strings:
+            return
+        node_ids = []
+        tasks_proxies = {}
+        for t_id, tproxy in list(self.task_proxies.items()):
+            if tproxy.cycle_point in point_strings:
+                node_ids.append(tproxy.id)
+                name, _ = TaskID.split(t_id)
+                tasks_proxies.setdefault(name, []).append(tproxy.id)
+                del self.task_proxies[t_id]
+        for name, id_list in tasks_proxies.items():
+            self.tasks[name].proxies[:] = [
+                tp for tp in self.tasks[name].proxies
+                if tp not in id_list]
+
+        families_proxies = {}
+        for f_id, fproxy in list(self.family_proxies.items()):
+            if fproxy.cycle_point in point_strings:
+                name, _ = TaskID.split(f_id)
+                families_proxies.setdefault(name, []).append(fproxy.id)
+                del self.family_proxies[f_id]
+        for name, id_list in families_proxies.items():
+            self.families[name].proxies[:] = [
+                fp for fp in self.families[name].proxies
+                if fp not in id_list]
+
+        g_eids = []
+        for e_id, edge in list(self.edges.items()):
+            if edge.source in node_ids:
+                del self.edges[e_id]
```

Comment:
There is a lot of `del` logic here, is it necessary to directly `del` the variable or would it be fine to use `list.remove` and let garbage collection do its job?

If so this can be simplified a bit which should make it a bit faster e.g:

```python
self.graph_edges = [
    edge.id
    for edge in self.edges.values()
    if edge.source in node_ids
]
```

Reply:
Not sure I follow... The loop does two things; `del` dictionary items if source id is in the list, else add the edge id to a list (for graph edges assignment). 

Are you suggesting I re-construct the dictionary with a dict comprehension and construct the graph edge.id list using a list comprehension separately?

Reply:
I think I must have missed that `self.edges` and `self.graph.edges` are different things.